src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests_RefReadOnly.cs

```diff
@@ -570,25 +986,580 @@ public class ReadOnlyAttribute : System.Attribute { }
+        {
+            var compilation = CreateCompilationWithMscorlib(@"
+public class Test
+{
+    void User()
+    {
+    }
+}");
+
+            compilation.VerifyEmitDiagnostics();
+            Assert.False(compilation.NeedsGeneratedIsReadOnlyAttribute);
+
+            var tree = compilation.SyntaxTrees.Single();
+            var model = compilation.GetSemanticModel(tree, ignoreAccessibility: false);
+
+            var userFunction = tree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Single(method => method.Identifier.Text == "User");
+            var position = userFunction.Body.CloseBraceToken.Position;
+            var localfunction = SyntaxFactory.ParseStatement("ref readonly int localFunction(int x) { return ref x; }");
+
+            Assert.True(model.TryGetSpeculativeSemanticModel(position, localfunction, out var newModel));
+            var localFunctionSymbol = newModel.GetDeclaredSymbol(localfunction);
+            Assert.NotNull(localFunctionSymbol);
+            Assert.False(compilation.NeedsGeneratedIsReadOnlyAttribute);
+        }
+
+        [Fact]
+        public void TryingPossibleBindingsForRefReadOnlyDoesNotPolluteCompilationForInvalidOnes()
+        {
+            var reference = CreateCompilationWithMscorlib(@"
+public delegate ref readonly int D1 ();
+public delegate ref int D2 ();
+").VerifyEmitDiagnostics();
+
+            Assert.True(reference.NeedsGeneratedIsReadOnlyAttribute);
+
+            var compilation = CreateCompilationWithMscorlibAndSystemCore(@"
+public class Test
+{
+    public void Process(D1 lambda, int x) { }
+    public void Process(D2 lambda, byte x) { }
+
+    void User()
+    {
+        byte byteVar = 0;
+        Process(() => { throw null; }, byteVar);
+    }
+}", references: new[] { reference.ToMetadataReference() });
+
+            compilation.VerifyEmitDiagnostics();
+            Assert.False(compilation.NeedsGeneratedIsReadOnlyAttribute);
+        }
+
+        [Fact]
+        public void RefReadOnlyErrorsForLambdasDoNotPolluteCompilationDeclarationsDiagnostics()
+        {
+            var reference = CreateCompilationWithMscorlib(@"
+public delegate int D (ref readonly int x);
+").EmitToImageReference();
+
+            var code = @"
+public class Test
+{
+    public void Process(D lambda) { }
+
+    void User()
+    {
+        Process((ref readonly int p) => p);
+    }
+}";
+
+            var compilation = CreateCompilationWithMscorlib(code, options: TestOptions.ReleaseModule, references: new[] { reference });
+
+            compilation.DeclarationDiagnostics.Verify();
+
+            compilation.VerifyDiagnostics(
+                // (10,35): error CS0518: Predefined type 'System.Runtime.CompilerServices.IsReadOnlyAttribute' is not defined or imported
+                //         Process((ref readonly int p) => p);
+                Diagnostic(ErrorCode.ERR_PredefinedTypeNotFound, "p").WithArguments("System.Runtime.CompilerServices.IsReadOnlyAttribute"));
```

Comment:
@AlekseyTs it might be related to #19353. will fix it here for now.

---
In reply to: [115352546](https://github.com/dotnet/roslyn/pull/18715#discussion_r115352546) [](ancestors = 115352546)